context/src/main/java/io/grpc/Context.java

```diff
@@ -228,25 +229,25 @@ private Context(Context parent, Object[][] keyValueEntries, boolean isCancellabl
 
   /**
    * Create a new context which is independently cancellable and also cascades cancellation from
-   * its parent. Callers should ensure that either {@link CancellableContext#cancel(Throwable)}
-   * or {@link CancellableContext#detachAndCancel(Context, Throwable)} are called to notify
-   * listeners and release the resources associated with them.
+   * its parent. Callers <em>must</em> ensure that either {@link
+   * CancellableContext#cancel(Throwable)} or {@link CancellableContext#detachAndCancel(Context,
+   * Throwable)} are called at a later point, in order to allow this context to be garbage
+   * collected.
    *
    * <p>Sample usage:
    * <pre>
    *   Context.CancellableContext withCancellation = Context.current().withCancellation();
    *   try {
-   *     executorService.execute(withCancellation.wrap(new Runnable() {
+   *     withCancellation.run(new Runnable() {
    *       public void run() {
    *         Context current = Context.current();
    *         while (!current.isCancelled()) {
    *           keepWorking();
    *         }
    *       }
    *     });
-   *     doSomethingRelatedWork();
-   *   } catch (Throwable t) {
-   *     withCancellation.cancel(t);
+   *   } finally {
+   *     withCancellation.cancel(null);
```

Comment:
Firstly, because that is the API and we can't change it. Secondly, because that would kill using it across threads or asynchronously. Thirdly, it would be quite subtle that if you call withValue(), that context wouldn't cancel on detach. We definitely don't want detach() to imply cancel() because they are commonly separate. In fact, gRPC's usage would break if detach() cancelled.